### DISCUSSION TITLE
Small setup controlled by ImSwitch

### DISCUSSION THREAD

**Discussion by jacopoabramo** [ImSwitch/ImSwitch#78] [General]
I'm currently trying to get a better knowledge of ImSwitch, and I'm considering doing a small test setup with a [pulse streamer ](https://www.swabianinstruments.com/pulse-streamer-8-2/) which will be connected to a laser in order to indirectly control it. The pulse streamer in question is the following:

As you can see the data connection is via Ethernet, which I'm not sure it's currently supported by ImSwitch. Here then are the following questions:

- would it make sense to integrate Ethernet as a low-level manager in order to communicate with the pulse streamer?
  - if this is feasible, what would be the procedure to integrate it in ImSwitch?
- if this would be overkill, is it sufficient to create a LaserManager device and integrate the Ethernet commands directly from there?

Thank you for your attention.

**Comment by xcasas** [comment]
Hi @jethro33, 

I think the easier way is to use the python library of Pulse Streamer, I see an example in “Client software” https://www.swabianinstruments.com/static/documentation/PulseStreamer/sections/gettingStarted.html .

 Then, I would say it requires the following steps:

1. Create a "SwabianLaserManager" or similar, you can use NidaqLaserManager as a template and reimplement the functions (setEnabled, setValue) using the python library of your device (import pulsestreamer). 

2. Create a new config file for testing your device, for example using example_coolLED.json as a template. Add your laser and in “managerName” add the name of the class you created “SwabianLaserManager” in this case. In managerProperties add anything that the python library needs (ethernet port or stuff automatically done so no need to do anything there. Following the example of CoolLEDLaserManager, the channel_index is retrieved as self.__channel_index = laserInfo.managerProperties['channel_index'] (line 27).

3. Then when selecting hardware setup you select your json file, either in “ Tools” -> “ Pick hardware setup” or in another file in the config files folder.
